File: src/main/java/journeymap/client/webmap/routes/Properties.java
package journeymap.client.webmap.routes;

import io.javalin.http.ContentType;
import io.javalin.http.Context;
import journeymap.client.JourneymapClient;
import journeymap.client.properties.WebMapProperties;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

public class Properties
{
    private static WebMapProperties webMapProperties = JourneymapClient.getWebMapProperties();
    private static Map<String, AtomicBoolean> propertiesMap = null;

    public static void get(Context ctx)
    {
        ctx.contentType(ContentType.APPLICATION_JSON);
        ctx.result(JourneymapClient.getWebMapProperties().toJsonString());
    }

    public static void post(Context ctx)
    {
        if (propertiesMap == null || propertiesMap.isEmpty())
        {
            WebMapProperties properties = JourneymapClient.getWebMapProperties();
            Map<String, AtomicBoolean> propMap = new HashMap<>();
            propMap.put("showGrid", properties.showGrid);
            propMap.put("showSelf", properties.showSelf);
            propMap.put("showWaypoints", properties.showWaypoints);
            propertiesMap = propMap;
        }
        for (String key : ctx.queryParamMap().keySet())
        {
            if (propertiesMap.containsKey(key))
            {
                AtomicBoolean property = propertiesMap.get(key);
                if (property == null)
                {
                    throw new NullPointerException("Properties value for " + key + " is null");
                }
                property.set(Boolean.parseBoolean(ctx.queryParam(key)));
            }
        }
        webMapProperties.save();
    }
}

File: src/main/kotlin/journeymap/client/service/webmap/kotlin/routes/properties.kt
package journeymap.client.service.webmap.kotlin.routes

import journeymap.client.JourneymapClient
import journeymap.client.properties.WebMapProperties
import spark.kotlin.RouteHandler
import java.util.concurrent.atomic.AtomicBoolean


val webMapProperties: WebMapProperties = JourneymapClient.getWebMapProperties()
var propertiesMap: Map<String, AtomicBoolean>? = null


internal fun propertiesGet(handler: RouteHandler): Any
{
    handler.response.raw().contentType = "application/json"
    return JourneymapClient.getWebMapProperties().toJsonString()
}


internal fun propertiesPost(handler: RouteHandler): Any
{
    if (propertiesMap == null || propertiesMap!!.isEmpty())
    {
        val properties: WebMapProperties = JourneymapClient.getWebMapProperties()
        val propMap = mutableMapOf<String, AtomicBoolean>()

        propMap["showGrid"] = properties.showGrid
        propMap["showSelf"] = properties.showSelf
        propMap["showWaypoints"] = properties.showWaypoints

        propertiesMap = propMap.toMap()
    }

    for (key in handler.queryMap().toMap().keys)
    {
        if (key in propertiesMap!!)
        {
            (
                    propertiesMap!![key] ?: error("Properties value for $key is null")
                    ).set(handler.queryMap(key).booleanValue())
        }
    }

    return webMapProperties.save()
}
